clang/lib/Sema/SemaChecking.cpp

```diff
@@ -2450,6 +2450,26 @@ static bool BuiltinCountZeroBitsGeneric(Sema &S, CallExpr *TheCall) {
   return false;
 }
 
+// The argument must be a class or struct with a member
+// named type.
+static bool CheckBuiltinSyclKernelName(Sema &S, CallExpr *TheCall) {
+  QualType ArgTy = TheCall->getArg(0)->getType();
+  const auto *RT = ArgTy->getAs<RecordType>();
```

Comment:
The primary intent of this function appears to be to validate that the type that identifies the kernel meets specific requirements. This is useful and is separable from the fact that the type is obtained by querying the type of the first argument in a call expression; I think it would be clearer to colocate the retrieval of the type with the code that checks for the number of arguments below.
```suggestion
// The SYCL kernel builtin functions identify the kernel to operate on
// by the type of one of the arguments. That type is required to be a
// class or structure type with a member typedef or type alias named
// "type".
static bool CheckBuiltinSyclKernelIDType(Sema &S, QualType T) {
  const auto *RT = T->getAs<RecordType>();
```